xapian-core/weight/tfidfweight.cc

```diff
@@ -211,42 +372,44 @@ TfIdfWeight::get_wdfn(Xapian::termcount wdf, Xapian::termcount doclen,
 	    return num / den;
 	}
 	default:
-	    AssertEq(c, 'n');
+	    AssertEq(wdf_norm_, WDF_NORM::NONE);
 	    return wdf;
     }
 }
 
 double
-TfIdfWeight::get_idfn(char c) const
+TfIdfWeight::get_idfn(IDF_NORM idf_norm_) const
 {
     Xapian::doccount termfreq = 1;
-    if (c != 'n') termfreq = get_termfreq();
+    if (idf_norm_ != IDF_NORM::NONE) termfreq = get_termfreq();
     double N = 1.0;
-    if (c != 'n' && c != 'f') N = get_collection_size();
-    switch (c) {
-	case 'n':
+    if ((idf_norm_ == IDF_NORM::PROB) || (idf_norm_ == IDF_NORM::SQUARE) ||
+	(idf_norm_ == IDF_NORM::TFIDF) || (idf_norm_ == IDF_NORM::PIVOTED))
+	N = get_collection_size();
+    switch (idf_norm_) {
+	case IDF_NORM::NONE:
 	    return 1.0;
-	case 'p':
+	case IDF_NORM::PROB:
 	    // All documents are indexed by the term
 	    if (N == termfreq) return 0;
 	    return log((N - termfreq) / termfreq);
-	case 'f':
+	case IDF_NORM::FREQ:
 	    return (1.0 / termfreq);
-	case 's':
+	case IDF_NORM::SQUARE:
 	    return pow(log(N / termfreq), 2.0);
-	case 'P':
+	case IDF_NORM::PIVOTED:
 	    return log((N + 1) / termfreq);
 	default:
-	    AssertEq(c, 't');
+	    AssertEq(idf_norm_, IDF_NORM::TFIDF);
```

Comment:
Since this is an enum, you can change the `default` into a specific case (nothing else will appear). Some compilers will warn if you don't provide cases for every possibility of an enum. Then you won't have to assert, either.

Reply:
When I removed default and added a case, I got this error -
error: control reaches end of non-void function [-Werror=return-type]

Reply:
You could make the final `case` in the `switch` just do `break;` and then have its `return` at the end of the function - then you won't get the "control reaches end of non-void function".

As James says, the major advantage of having all the enum values listed as explicit cases with no `default:` is that compilers will warn if a new enum value is added without new cases being added for it (which is an easy mistake for a future developer working on the code to make).  If there's a `default:` then there's no such warning (since the new value is handled by the `default:`).